File: benchmarks/python/pipeline.py
# pipeline — see ../praxis/pipeline.px for the description.
#
# Generator expressions are Python's fused pipeline: no intermediate list is
# built, which is the same property §6.3 claims for a Praxis chain.
# `functools.reduce` is the `fold`.
import sys
from functools import reduce


def main():
    n = int(sys.stdin.read().strip())

    v = []
    i = 0
    while i < n:
        v.append((i * 7919 + 13) % 1000003)
        i += 1

    acc = 0
    r = 0
    while r < 12:
        salt = r
        acc = (acc + sum(y for y in ((x * 3 + salt) % 1000003 for x in v) if y % 5 != 0)) % 1000003
        r += 1

    folded = 0
    s = 0
    while s < 12:
        salt = s
        folded = (
            folded
            + reduce(
                lambda a, x: (a * 2 + x) % 1000003,
                ((x + idx * salt) % 1000003 for idx, x in enumerate(v)),
                0,
            )
        ) % 1000003
        s += 1

    print(len(v))
    print(acc)
    print(folded)


main()

File: benchmarks/rust/pipeline.rs
// pipeline — see ../praxis/pipeline.px for the description.
use std::io::Read;

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).unwrap();
    let n: i64 = buf.trim().parse().unwrap();

    let mut v: Vec<i64> = Vec::new();
    let mut i: i64 = 0;
    while i < n {
        v.push((i * 7919 + 13) % 1000003);
        i += 1;
    }

    let mut acc: i64 = 0;
    let mut r: i64 = 0;
    while r < 12 {
        let salt = r;
        acc = (acc
            + v.iter()
                .map(|x| (x * 3 + salt) % 1000003)
                .filter(|x| x % 5 != 0)
                .sum::<i64>())
            % 1000003;
        r += 1;
    }

    let mut folded: i64 = 0;
    let mut s: i64 = 0;
    while s < 12 {
        let salt = s;
        folded = (folded
            + v.iter()
                .enumerate()
                .map(|(idx, x)| (x + idx as i64 * salt) % 1000003)
                .fold(0i64, |a, x| (a * 2 + x) % 1000003))
            % 1000003;
        s += 1;
    }

    println!("{}", v.len());
    println!("{acc}");
    println!("{folded}");
}
